core/src/types/blocking_read/buffer_iterator.rs

```diff
@@ -0,0 +1,98 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use bytes::Buf;
+use rayon::prelude::*;
+
+use crate::raw::*;
+use crate::*;
+
+pub(super) struct BufferIterator {
+    inner: Arc<dyn oio::BlockingRead>,
+    chunk: Option<usize>,
+
+    offset: u64,
+    end: Option<u64>,
+    concurrent: usize,
+    finished: Arc<AtomicBool>,
+}
+
+impl BufferIterator {
+    pub fn new(
+        inner: Arc<dyn oio::BlockingRead>,
+        chunk: Option<usize>,
+        concurrent: usize,
+        offset: u64,
+        end: Option<u64>,
+    ) -> Self {
+        Self {
+            inner,
+            chunk,
+            offset,
+            end,
+            concurrent,
+            finished: Arc::new(AtomicBool::new(false)),
+        }
+    }
+}
+
+impl Iterator for BufferIterator {
+    type Item = Result<Buffer>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.end.unwrap_or(u64::MAX) {
+            return None;
+        }
+        if self.finished.load(Ordering::Relaxed) {
+            return None;
+        }
+
+        let mut bufs = Vec::with_capacity(self.concurrent);
+        let interval_size = self.chunk.unwrap_or(4 * 1024 * 1024) as u64;
+
+        let intervals: Vec<(u64, u64)> = (0..self.concurrent as u64)
```

Comment:
Please don't calcluate ranges at once.

We can build a `RangeIterator` that only generate the next range to read. `BufferIterator` stores `RangeIterator`, every time users call `next`, `BufferIterator` fetchs a new range and than read as a new buffer.

Reply:
> Please don't calcluate ranges at once.
How can this be paralleled?

Reply:
> How can this be paralleled?

After re-visit the rayon API docs, I found that we can't call `next` on a `ParallelIterator` directly. So we can change the `RangeIterator` to return a `ParallelIterator` instead. 

Every time users call `next` on `BufferIterator`, we build a new `ParallelIterator`, extend into buffer and yield from it.

The whole workflow looks like:

```rust
RangeIterator::next() -> Range<u64>


struct BufferIterator {
  it: RangeIterator,
  buf: Vec<Buffer>
}

impl Iterator for BufferIterator {
    type Item = Result<Buffer>;

    fn next(&mut self) -> Option<Self::Item> {
        // fecth from buffer first.
        // Build a new ParallelIterator and collect it.
        // Return None if ranges have been all consumed.
    }
}
```